Apache Ignite File System

dmitrievanthony opened the pull request:
This is a proposal to support [Apache Ignite File System](https://ignite.apache.org/features/igfs.html) in TensorFlow.

[Apache Ignite](https://ignite.apache.org/) is a memory-centric distributed database, caching, and processing platform for transactional, analytical, and streaming workloads, delivering in-memory speeds at petabyte scale. This proposal is a part of a more global initiative to implement so called "TensorFlow on Apache Ignite" ([IGNITE-8335](https://issues.apache.org/jira/browse/IGNITE-8335), [Design Document](https://docs.google.com/document/d/1jROIahK1rc7bSgOvhJhfpMqIGvht_IE8zn5NAt6x8ks)).

The integration is based on [custom filesystem plugin](https://www.tensorflow.org/extend/add_filesys) from TensorFlow side and IGFS Native API from Apache Ignite side. More information about this module you can find in [README.md](https://github.com/dmitrievanthony/tensorflow/blob/apache-ignite-igfs/tensorflow/contrib/igfs/README.md).

Tests have also been added. They use docker to hide configuration complexity, so that the implemented functionality can be tested quite simply by manual run.

**Be aware that this pull request is based on previous #22210 that is not merged yet (significant part of the code you see here will be merged in #22210).**

dmitrievanthony committed:
Add IGFS (Apache Ignite File System) support.

artemmalykh committed:
Add IGFS (Apache Ignite File System) support.

yifeif commented:
Thanks @dmitrievanthony. Could you check the cla for all contributors who have commits in this PR?

dmitrievanthony commented:
Hi, @yifeif. Thank you for attention. Did I understand correctly from the message above says that all authors have signed CLA and the problem is only that the committer and the author of one commit are not the same? If so, I asked @artemmalykh to write here that he's ok with this PR.

artemmalykh commented:
I'm totally OK with PR!

yifeif commented:
Thanks @dmitrievanthony and @artemmalykh. Looks good now.

dmitrievanthony commented:
@yifeif, I just rebased this branch, so commits haven't been changed. I think CLA check should be fine. I still waiting #22210 to be merged so I'll make one more rebase when it will be merged and ping you again.

dmitrievanthony committed:
Update after review of #22210.

dmitrievanthony commented:
Hi, @mrry, @martinwicke, @yifeif.

After merge of #22210 I rebased and updated this pull request, so now it's ready for review. It's quite big, but less that Ignite Dataset. 

Derek, I'll be grateful if you have a look and start reviewing. Please let me know if changes required.

Would be great if someone starts tests and clicks `cla:yes` (cla checker is too suspicious in case of rebases).

dmitrievanthony commented:
Hi, @mrry. Any updates?

mrry commented on tensorflow/contrib/igfs/kernels/igfs_client.h:
```diff
@@ -0,0 +1,109 @@
+             const string &user_name);
+  ~IGFSClient();
+
+  inline Status Handshake(CtrlResponse<HandshakeResponse> *res) {
```
`inline` is implied for member functions defined in the class definition.

dmitrievanthony replied:
Yes, agree. It's exactly member function defined in the class definition.

dmitrievanthony replied:
Okay, I got it. I've removed `inline` in headers here and in `ignite_client.h` and `ignite_byte_swapper.h` as well.

mrry commented on tensorflow/contrib/BUILD:
```diff
@@ -126,6 +126,7 @@ py_library(
     }) + select({
         "//tensorflow:with_ignite_support": [
             "//tensorflow/contrib/ignite",
+            "//tensorflow/contrib/igfs",
```
Can we combine these two modules into one? IGFS seems like it would fit under `"//tensorflow/contrib/ignite"`.

dmitrievanthony replied:
I would prefer to have them in separate modules by several reasons. First of all, they have very different functionality (`"//tensorflow/contrib/ignite"` provides an ability to read data from Apache Ignite Cache while `"//tensorflow/contrib/igfs"` provides allows to use file system, based on Apache Ignite). User may want to use only one feature, but not other. It looks like very reasonable in some cases. I think it's a case when "write programs that do one thing and do it well" is good, isn't it? And by the way, after `contrib` sunsetting it makes even more sense to have them in separate packages.

Are you agree?

dmitrievanthony replied:
Sounds reasonable. I merged two modules together.

mrry commented on tensorflow/contrib/igfs/kernels/igfs.cc:
```diff
@@ -0,0 +1,319 @@
+  return Status::OK();
+}
+
+std::shared_ptr<IGFSClient> IGFS::CreateClient() const {
```
Is the "shared"-ness ever used here? It looks like this could return a `unique_ptr`.

dmitrievanthony replied:
Replaced by `unique_ptr`, done.

mrry commented on tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc:
```diff
@@ -0,0 +1,141 @@
+    TF_RETURN_IF_ERROR(ReadString(&key));
+    TF_RETURN_IF_ERROR(ReadString(&val));
+
+    res->insert(std::pair<string, string>(key, val));
```
You could `std::move(key), std::move(val)` here to avoid copying the strings again.

dmitrievanthony replied:
Done.

mrry commented on tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc:
```diff
@@ -0,0 +1,141 @@
+Status ExtendedTCPClient::WriteString(string str) {
+  if (!str.empty()) {
+    TF_RETURN_IF_ERROR(WriteBool(false));
+    unsigned short l = str.length();
```
This is an unsafe cast and should raise a compiler warning: `str.length()` returns a `size_t`. You need to check that it fits in the range of an `unsigned short` before casting.

dmitrievanthony replied:
Done.

mrry commented on tensorflow/contrib/igfs/kernels/igfs_messages.cc:
```diff
@@ -0,0 +1,344 @@
+                                 const std::map<string, string> &properties)
+    : Request(command_id_),
+      user_name_(std::move(user_name)),
+      path_(std::move(path)),
```
`path` is a `const string&`, so moving it has no effect. If you want to avoid a copy, you need to take `path` as a (by-value) `string` or a `string&&`.

dmitrievanthony replied:
Okay, just removed `copy()`.

mrry commented on tensorflow/contrib/igfs/kernels/igfs_messages.h:
```diff
@@ -0,0 +1,356 @@
+    int32_t len;
+    TF_RETURN_IF_ERROR(client->ReadInt(&len));
+
+    entries = std::vector<T>();
```
Suggestion: `entries.clear()` is slightly more readable, since the object must already have been constructed.

dmitrievanthony replied:
Done.

dmitrievanthony committed:
Update Apache Ignite IGFS after review.

dmitrievanthony commented on tensorflow/contrib/igfs/__init__.py:
```diff
@@ -0,0 +1,36 @@
+plugged into Hadoop or Spark deployments. This contrib package contains an
+intergration between IGFS and TensorFlow.
+
+@@IGFS
```
Removed.

dmitrievanthony commented on tensorflow/contrib/igfs/kernels/igfs.cc:
```diff
@@ -0,0 +1,319 @@
+
+namespace tensorflow {
+
+string GetEnvOrElse(const string &env, string default_value) {
```
Done, marked as `static`.

dmitrievanthony committed:
Update Apache Ignite IGFS after review.

dmitrievanthony commented:
Hi, @mrry. Thanks for review. I updated code in accordance with all your comments, please have a look.

dmitrievanthony commented:
Hi @mrry. Any updates?

dmitrievanthony commented:
Hi, @mrry, @yifeif. I don't clearly understand current state. Do you have some concerns about this PR or it's ready to be merged?

dmitrievanthony committed:
Update Apache Ignite IGFS after review.

dmitrievanthony commented on tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.cc:
```diff
@@ -0,0 +1,143 @@
+  if (!str.empty()) {
+    TF_RETURN_IF_ERROR(WriteBool(false));
+    size_t l = str.length();
+    if (l > 0xFFFF) return errors::InvalidArgument("String is too long");
```
Good point, done.

dmitrievanthony commented:
Hi @mrry. I merged `ignite` and `igfs` modules together and fixed other remarks. Please have a look.

mrry commented on tensorflow/contrib/ignite/BUILD:
```diff
@@ -31,39 +32,88 @@ tf_custom_op_library(
+        "kernels/igfs/igfs_writable_file.h",
+    ],
+    deps = [
+        "//tensorflow/contrib/ignite:ignite_client",
```
Nit: This should be `":ignite_client"` now.

dmitrievanthony replied:
Done.

mrry commented on tensorflow/contrib/ignite/BUILD:
```diff
@@ -31,39 +32,88 @@ tf_custom_op_library(
     deps = [":dataset_kernels"],
 )
 
+tf_custom_op_library(
```
Since `":igfs_kernels"` and `":dataset_kernels"` share some symbols (in `":ignite_client"`), to avoid the ODR issues, it'll be easiest to merge "_dataset_ops.so" and "_igfs_ops.so" into a single shared object (and hence `tf_custom_op_library()` rule).

dmitrievanthony replied:
Okay, good point. I've merged them into a single shared library `_ignite_ops.so`.

dmitrievanthony committed:
Update Apache Ignite IGFS after review.

dmitrievanthony commented:
Hi @mrry. I've updated the code, please have a look.

dmitrievanthony committed:
Remove unused dependencies.

dmitrievanthony commented:
It looks like Sanity failed because of `//tensorflow/core:lib` dependency in `igfs_kernels`. I've removed it, so please rerun tests.

dmitrievanthony commented:
Thanks @mrry, looks like everything is fixed. Currently only one test fails and it fails in non related to my code part.

Hope the PR will be merged soon :)
